### ISSUE TITLE
Cannot import ESM modules when using TypeScript

### ISSUE THREAD

**Issue by vvo** [knex/knex#4447]
# Environment

Knex version: 0.95.4
Database + version: sqlite3 5.0.2
OS: macOS Big Sur 11.1

@lorefnon (since related to TypeScript)

Hi there, I spent quite a bunch of time on this without finding the proper way to resolve it.

# Bug

1. Explain what kind of behavior you are getting and how you think it should do

When trying to `import randomInt from "random-int"` from a TypeScript migration file we get an error.

2. Error message

```txt
Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /Users/[redacted]/knex-typescript-migrations-esm/node_modules/random-int/index.js
require() of ES modules is not supported.
require() of /Users/[redacted]/knex-typescript-migrations-esm/node_modules/random-int/index.js from /Users/[redacted]/knex-typescript-migrations-esm/migrations/20210419131555_test.ts is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /Users/[redacted]/knex-typescript-migrations-esm/node_modules/random-int/package.json.

    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1080:13)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/Users/[redacted]/knex-typescript-migrations-esm/migrations/20210419131555_test.ts:2:1)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Module.m._compile (/Users/[redacted]/knex-typescript-migrations-esm/node_modules/ts-node/src/index.ts:1056:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/[redacted]/knex-typescript-migrations-esm/node_modules/ts-node/src/index.ts:1059:12)
```

Running with --esm sort of works but breaks TypeScript:

```txt
export async function up(knex: Knex): Promise<void> {
                             ^

SyntaxError: Invalid or unexpected token
    at Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
```

3. Reduced test code: https://github.com/vvo/knex-typescript-migrations-esm

Thanks!

**Comment by cayter** [comment]
I looked into the source code and found:

- the migrator is relying on [this](https://github.com/knex/knex/blob/master/lib/migrations/util/import-file.js#L10) to decide if it should use `import()` to load the `.ts` file.
- however, the [function](https://github.com/knex/knex/blob/master/lib/migrations/util/is-module-type.js#L3) it's relying on only works with `.mjs` file extension or we would need to run `npm_package_json=package.json knex migrate:latest` with `"type": "module"` configured in the `package.json`.

I tested setting `npm_package_json` approach and it worked.

**Comment by cayter** [comment]
I looked into the source code and found:

- the migrator is relying on [this](https://github.com/knex/knex/blob/master/lib/migrations/util/import-file.js#L10) to decide if it should use `import()` to load the `.ts` file.
- however, the [function](https://github.com/knex/knex/blob/master/lib/migrations/util/is-module-type.js#L3) it's relying on only works with `.mjs` file extension or we would need to run `npm_package_json=package.json knex migrate:latest` with `"type": "module"` configured in the `package.json`.

I tested by setting `npm_package_json` approach and it worked. If you're interested in fixing it and submit a PR, it's as easy as fixing https://github.com/knex/knex/blob/master/lib/migrations/util/is-module-type.js to also check filepath that ends with `.ts`:

```js
const { readFile } = require('./fs');

module.exports = async function isModuleType(filepath) {
  if (process.env.npm_package_json) {
    // npm >= 7.0.0
    const packageJson = JSON.parse(
      await readFile(process.env.npm_package_json, 'utf-8')
    );
    if (packageJson.type === 'module') {
      return true;
    }
  }
  return process.env.npm_package_type === 'module' || filepath.endsWith('.mjs') || filepath.endsWith('.ts');
};
```

However, I think the code above isn't solid enough. By right, it should check if the filepath ends with any value configured in `MigratorConfig.extensions` which is the single source of truth for our migrator configuration.

**Comment by kibertoad** [comment]
@cayter Thank you for looking into it! Would you consider sending a PR for this?

**Comment by azban** [comment]
not sure this is totally relevant, but seems like https://github.com/cfware/get-package-type#readme may be a robust alternative to using npm_package_json.. especially given that the yarn 2 core dev(s?) seem opposed to adding that variable. as a result, i'm guessing this is currently broken with yarn 2.

see:
- https://github.com/yarnpkg/berry/issues/1659#issuecomment-770465913
- https://github.com/yarnpkg/berry/issues/2436

**Comment by lehni** [comment]
I've created a PR for this now, I hope this solves it for everybody.

**Comment by lehni** [comment]
With #5071 merged (81b7edd115a15849f54d5176ce4720fc5da34314), I believe this issue can be closed now, cc @kibertoad

**Comment by kibertoad** [comment]
Released in 1.0.5

**Comment by andrewchilds** [comment]
Hi @lehni and @kibertoad! After upgrading Knex from 1.0.2 to 1.0.7, I was having trouble running migrations locally, and tracked it down to this PR. I use yarn, ESM, and `.js` file extensions in my project.

Here's the command I've been using:

```
node_modules/knex/bin/cli.js migrate:up --esm --knexfile ./path/to/knexfile.js
```

Knex would evaluate the next migration file, but not actually execute the `up` command. I finally realized that by removing the `--esm` flag, the migration would run. This seems at odds with the knexjs documentation at https://knexjs.org/#esm-interop. What is the behavior of `--esm` flag in Knex 1.0.5 and newer? I'm happy to help update the documentation, but not sure what the new behavior is.

btw - @lehni I've been a fan of your work for many years! And @kibertoad, thank you for maintaining this extremely useful library.
